ecl/hql/hqlgram2.cpp

```diff
@@ -3346,6 +3358,147 @@ unsigned HqlGram::getExtraLookupFlags(IHqlScope * scope)
     return 0;
 }
 
+IHqlExpression * HqlGram::lookupNearestSymbol(IIdAtom * searchName)
+{
+    NearestSymbol nearest(searchName);
+    lookupNearestSymbol(searchName, nearest);
+    return LINK(nearest.querySymbol());
+}
+
+IHqlExpression * HqlGram::lookupNearestSymbol(IIdAtom * searchName, NearestSymbol & nearest)
+{
+    //Check periodically if parsing a referenced identifier has caused the compile to abort.
+    if (lookupCtx.isAborting())
+        aborting = true;
+
+    unsigned penalty = 0;
+    try
+    {
+        // If there is a temporary scope, we only look up in that (and it must exist!).
+        if (dotScope)
+        {
+            if (dotScope->getOperator() == no_enum)
+            {
+                NearestSymbol tempNearest(searchName);
```

Comment:
The use of all these temporary NearestSymbol objects leads me to suspect that the logic isn't quite right.  Should the penalty be being passed in to the lookupNearestSymbol function instead?  When should the penalty be incremented?  

How should the penalty decide on the best match?  You add it to the distance - but should it instead work more as a tie breaker?  If the penalty does affect the distance, then it should be passed in to the lookup function because it means the best match can be short-circuited more quickly.